tests/miral/render_scene_into_surface.cpp

```diff
@@ -0,0 +1,128 @@
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "render_scene_into_surface.h"
+
+#include <miral/test_server.h>
+#include <gtest/gtest.h>
+#include <gmock/gmock.h>
+
+#include "mir/server.h"
+#include "mir/compositor/scene.h"
+#include "mir/compositor/scene_element.h"
+#include "mir/graphics/renderable.h"
+#include "mir/scene/surface.h"
+
+namespace mtf = mir_test_framework;
+
+using namespace testing;
+using namespace miral;
+
+
+class RenderSceneIntoSurfaceTest : public TestServer
+{
+public:
+    RenderSceneIntoSurfaceTest()
+    {
+        start_server_in_setup = false;
+        add_to_environment("MIR_SERVER_PLATFORM_DISPLAY_LIBS", "mir:virtual");
+        add_to_environment("MIR_SERVER_VIRTUAL_OUTPUT", "800x600");
+    }
+};
+
+TEST_F(RenderSceneIntoSurfaceTest, set_capture_area_before_starting)
+{
+    RenderSceneIntoSurface render_scene_into_surface;
+    add_server_init(render_scene_into_surface);
+    render_scene_into_surface.capture_area(Rectangle(
+        Point(100, 100),
+        Size(200, 200)
+    ));
+
+    start_server();
+    auto const scene = server().the_scene();
+    auto const elements = scene->scene_elements_for(this);
+    EXPECT_THAT(elements.size(), Eq(1));
+    auto const& element = elements.at(0);
+    EXPECT_THAT(element->renderable()->screen_position(), Eq(Rectangle(
+        Point(0, 0),
+        Size(200, 200)
+    )));
+}
+
+TEST_F(RenderSceneIntoSurfaceTest, set_capture_area_after_starting)
+{
+    RenderSceneIntoSurface render_scene_into_surface;
+    add_server_init(render_scene_into_surface);
+
+    start_server();
+    render_scene_into_surface.capture_area(Rectangle(
+        Point(100, 100),
+        Size(200, 200)
+    ));
+    auto const scene = server().the_scene();
+    auto const elements = scene->scene_elements_for(this);
+    EXPECT_THAT(elements.size(), Eq(1));
+    auto const& element = elements.at(0);
+    EXPECT_THAT(element->renderable()->screen_position(), Eq(Rectangle(
+        Point(100, 100),
+        Size(200, 200)
+    )));
+}
+
+TEST_F(RenderSceneIntoSurfaceTest, set_overlay_cursor_before_starting)
+{
+    RenderSceneIntoSurface render_scene_into_surface;
+    render_scene_into_surface.overlay_cursor(true);
+    add_server_init(render_scene_into_surface);
+    start_server();
+}
+
+TEST_F(RenderSceneIntoSurfaceTest, set_overlay_cursor_after_starting)
+{
+    RenderSceneIntoSurface render_scene_into_surface;
```

Comment:
ASAN is correct. This scope and the variables in it are lost following the `start_server()` call. There is nothing to ensure that they are not accessed following scope exit.

It looks like `render_scene_into_surface` could be moved into the test fixture (which fixes the lifetime) and `add_server_init(render_scene_into_surface);` could be called in `SetUp()` (which removes some setup from all the tests)